https://issues.redhat.com/browse/ACM-26213 note about fail-over (#8417)
open-cluster-management/rhacm-docs: governance/gatekeeper_operator/config_gk_operator.adoc

Install the {gate} from the {olm} catalog to install Gatekeeper on your cluster. With {acm-short} you can use a policy to install the {gate} by using the governance framework. After you install the {gate}, configure the {gate} custom resource to install Gatekeeper.

*Required access*: Cluster administrator

.Prerequisites

- You need a {acm} or {ocp} Plus subscription to install Gatekeeper and apply Gatekeeper policies to your cluster.
- Understand how to use the Operator Lifecycle Manager and the {ocp-catalog} by completing the _Adding Operators to a cluster_ and the _Additional resources_ section in the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.18/html/operators/administrator-tasks#olm-adding-operators-to-a-cluster[{ocp-short} documentation].

.Procedure

The {gate} custom resource tells the {gate} to start the Gatekeeper installation on the cluster. Complete the following steps to configure the {gate} custom resource:

> 

Reviewer: Could we maybe start this section with the most basic configuration, and then the remaining YAML can be snippets, something like:

>The {gate} custom resource is a cluster-scoped resource and must be named `gatekeeper`. Both the validating and mutating webhooks are enabled by default. Use the `validatingWebhook` and `mutatingWebhook` parameters to enable or disable the webhooks. See the following example:
>
>```yaml
>apiVersion: operator.gatekeeper.sh/v1alpha1
>kind: Gatekeeper
>metadata:
>  name: gatekeeper
>spec:
>  validatingWebhook: Enabled
>  mutatingWebhook: Enabled
>```

Author: Should this be a part of the procedure or come before the start? As of now, I added it as a step

Reviewer: Yes, I think this looks good, thanks!

> Should this be a part of the procedure or come before the start? As of now, I added it as a step

. Enable the `auditEventsInvolvedNamespace` parameter to manage the namespace audit event you want to create. When you enable this parameter, the Gatekeeper controller deployment runs with the following argument: `--audit-events-involved-namespace=true`. See the following example:

+
[source,yaml]
----
apiVersion: operator.gatekeeper.sh/v1alpha1
kind: Gatekeeper
metadata:
  name: gatekeeper
spec:
  audit:
    auditChunkSize: 66
    auditEventsInvolvedNamespace: Enabled
    auditFromCache: Enabled
    auditInterval: 10s
    constraintViolationLimit: 55
---

> ----

Reviewer: With the leading basic YAML in place, I think we can trim this way down to just the breadcrumbs?

```asciidoc
[source,yaml]
----
spec:
  audit:
    auditEventsInvolvedNamespace: Enabled
----
```

. Specify `containerArguments` in the `audit` and `webhook` specification by providing a list of argument names and values to pass to the container. Omit leading dashes from the argument name. An omitted value is treated as `true`. Arguments that you provide are ignored if the argument is set previously by the operator or configurations from other fields. See the following list of flags that are deny-listed and are not currently supported:

+
- `port`
- `prometheus-port`
- `health-addr`
- `validating-webhook-configuration-name`
- `mutating-webhook-configuration-name`
- `disable-cert-rotation`
- `client-cert-name`
- `tls-min-version`

+
View the following partial YAML file example:

+
[source,yaml]
----
    containerArguments:
    - name: ""
      value: ""
    resources:
      limits:
        cpu: 500m
        memory: 150Mi
      requests:
        cpu: 500m
        memory: 130Mi
---

> ----

Reviewer: I think this could be a better example than the empty values. It looks like there's a link with a nice list of arguments we can point to also: https://open-policy-agent.github.io/gatekeeper/website/docs/runtime-flags

```asciidoc
[source,yaml]
----
spec:
  webhook:
    containerArguments:
    - name: exempt-namespace-prefix
      value: sample-test-
----
```

> . For Gatekeeper version 3.20 and later, specify `audit.podAnnotations` that are specific to the audit pod by providing a list of annotation names and values to add to the pod. An omitted value is treated as `true`. View the following example:

Reviewer: The "true" thing doesn't apply to the annotations--just the `containerArguments`.

```suggestion
. For Gatekeeper version 3.20 and later, specify `audit.podAnnotations` that are specific to the audit pod by providing a list of annotation names and values to add to the pod. View the following example:
```

+
[source,yaml]
----
    emitAuditEvents: Enabled
    logLevel: DEBUG
    podAnnotations: {} 
    replicas: 1
---

> ----

Reviewer: Like previous, could be better to drop an example value in there.

```asciidoc
[source,yaml]
----
spec:
  audit:
    podAnnotations:
      foo: bar
----
```

. Enable the `admissionEventsInvolvedNamespace` parameter to manage the namespace admission event you want to create. When you enable this parameter, the Gatekeeper controller deployment runs with the following argument: `--admission-events-involved-namespace=true`. See the following example:

+
[source,yaml]
----
  validatingWebhook: Enabled
  mutatingWebhook: Enabled
  webhook:
    admissionEventsInvolvedNamespace: Enabled
---

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  webhook:
    admissionEventsInvolvedNamespace: Enabled
----
```

. *Optional:* If you want to enforce your Gatekeeper constraints, set the `failurePolicy` parameter to `Fail`. By default, the `failurePolicy` parameter for Gatekeeper is set to `Ignore`. If the Gatekeeper webhook becomes unavailable, constraints are not enforced.

+
[source,yaml]
----
    containerArguments:
    - name: ""
      value: ""
    disabledBuiltins:
     - http.send
    emitAdmissionEvents: Enabled
    failurePolicy: Fail
---

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  webhook:
    failurePolicy: Fail
----
```

> . Define `operations` for your webhook. Use one or more of the following supported values: `CREATE`, `UPDATE`, `CONNECT`, and `DELETE`. See the following example:

Reviewer: ```suggestion
. Define `operations` for your webhook. Use one or more of the following supported values: `CREATE`, `UPDATE`, `CONNECT`, and `DELETE`. By default, the operations are `CREATE` and `UPDATE`. See the following example:
```

+
[source,yaml]
----
     operations:
     - CREATE
     - UPDATE
     - CONNECT
    replicas: 3
    resources:
      limits:
        cpu: 480m
        memory: 140Mi
      requests:
        cpu: 400m
        memory: 120Mi
---

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  webhook:
     operations:
     - CREATE
     - UPDATE
     - CONNECT
----
```

> . For version 3.20 and later, specify `rules` to overwrite the `rules` configuration in the `mutatingWebhookConfig` specification. The `rules` parameter preceeds the `operations` field. See the following example:

Reviewer: ```suggestion
. For version 3.20 and later, specify `rules` to overwrite the `rules` configuration in the `webhook` and `mutatingWebhookConfig` specification. The `rules` parameter supersedes the `operations` field. See the following example:
```

+
[source,yaml]
----
    rules:
    - operations: []
      resources: []
    timeoutSeconds: 3
  ---

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  webhook:
    rules:
    - apiGroups:
      - ""
      apiVersions:
      - v1
      operations:
      - CREATE
      - UPDATE
      resources:
      - Pods
      - Deployments
----
```

> . For version 3.20 and later, specify the timeout for the webhook configuration by defining a value for the `timeoutSeconds` parameter.

Reviewer: ```suggestion
. For version 3.20 and later, specify a custom timeout for the webhook configuration by defining a value for the `timeoutSeconds` parameter.
```

+
[source,yaml]
----
    rules:
    - operations: []
      resources: []
    timeoutSeconds: 1

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  webhook:
    timeoutSeconds: 10
----
```

. For version 3.20 and later, configure the mutating webhook by adding configurations for the `mutatingWebhookConfig` specification. The configurations override the values from `spec.webhook` or are specific to the mutating webhook. See the following example:

+
[source,yaml]
----
  mutatingWebhookConfig:
    failurePolicy: ""
    logMutations: Disabled
    mutationAnnotations: Disabled
    namespaceSelector: {}
    operations: []
    rules:
---

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  mutatingWebhookConfig:
    failurePolicy: Ignore
    logMutations: Disabled
    mutationAnnotations: Disabled
    namespaceSelector: {}
    operations: []
    rules: []
    timeoutSeconds: 1
----
```

> . Use the `config` section to exclude namespaces from certain processes for all constraints on your cluster. See the following example:

Reviewer: If it might be helpful, this explains the difference between the Config resource and the flags for exempting namespaces: https://open-policy-agent.github.io/gatekeeper/website/docs/exempt-namespaces/

```suggestion
. Use the `config` section to exclude namespaces from certain processes for all constraints on your cluster using the Gatekeeper `Config` custom resource. See the following example:
```

+
[source,yaml]
----
  config:
    matches: 
      - excludedNamespaces: ["test-*", "my-namespace"] 
        processes: ["*"] 
---

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  config:
    matches: 
      - excludedNamespaces: ["test-*", "my-namespace"] 
        processes: ["*"] 
----
```

. The `disableDefaultMatches` parameter is a boolean parameter that disables appending the default exempt namespaces provided by the Gatekeeper operator. The default exempt namespaces are {ocp-short} or Kubernetes system namespaces. By default, this parameter is set to `false` to allow the default namespaces to be appended. See the following example:

+
[source,yaml]
----
  config:
    matches: 
      - excludedNamespaces: ["test-*", "my-namespace"] 
        processes: ["*"] 
    disableDefaultMatches: false
---

> ----

Reviewer: ```asciidoc
[source,yaml]
----
spec:
  config:
    disableDefaultMatches: false
----
```

.Procedure

The {gate} custom resource is a cluster-scoped resource and must be named `gatekeeper`. Both the validating and mutating webhooks are enabled by default. Use the `validatingWebhook` and `mutatingWebhook` parameters to enable or disable the webhooks. Complete the following steps to configure the {gate} custom resource:

. Enable the `validatingWebhook` and `mutatingWebhook` parameters. Your YAML file might resemble the following example:

+
[source,yaml]
----
apiVersion: operator.gatekeeper.sh/v1alpha1
kind: Gatekeeper
metadata:
 name: gatekeeper
spec:
 validatingWebhook: Enabled
 mutatingWebhook: Enabled
----

. Enable the `auditEventsInvolvedNamespace` parameter to manage the namespace audit event you want to create. When you enable this parameter, the Gatekeeper controller deployment runs with the following argument: `--audit-events-involved-namespace=true`. See the following example:

+
[source,yaml]
----
...
spec:
  audit:
    auditEventsInvolvedNamespace: Enabled

> ---

Reviewer: Is this extra set of dashes necessary?

```suggestion
```

. Specify `containerArguments` in the `audit` and `webhook` specification by providing a list of argument names and values to pass to the container. Omit leading dashes from the argument name. An omitted value is treated as `true`. Arguments that you provide are ignored if the argument is set previously by the operator or configurations from other fields. See the following list of flags that are deny-listed and are not currently supported:

+
- `port`
- `prometheus-port`
- `health-addr`
- `validating-webhook-configuration-name`
- `mutating-webhook-configuration-name`
- `disable-cert-rotation`
- `client-cert-name`
- `tls-min-version`

. For Gatekeeper version 3.20 and later, specify `audit.podAnnotations` that are specific to the audit pod by providing a list of annotation names and values to add to the pod. An omitted value is treated as `true`.

. Enable the `admissionEventsInvolvedNamespace` parameter to manage the namespace admission event you want to create. When you enable this parameter, the Gatekeeper controller deployment runs with the following argument: `--admission-events-involved-namespace=true`.

> . *Optional:* If you want to enforce your Gatekeeper constraints, set the `failurePolicy` parameter to `Fail`. By default, the `failurePolicy` parameter for Gatekeeper is set to `Ignore`. If the Gatekeeper webhook becomes unavailable, constraints are not enforced.

Author: Instead of adding the important note, I suggest we add it as a step. I created steps here to remove the callouts

Reviewer: If I recall, the removal of the callouts is because they're not supported in DITA, correct?

Would it make sense to move the YAML above the descriptions?

Author: You're correct about the call-outs not being supported in DITA. We can move the YAML before the steps or break the YAML into smaller pieces
